rasa/nlu/data_augmentation/augmenter.py

```diff
@@ -0,0 +1,623 @@
+import random
+from typing import Any, Dict, List, Set, Text, Tuple
+
+from rasa.model import get_model
+from rasa.shared.core.domain import Domain
+import rasa.shared.utils.io
+from rasa.shared.nlu.training_data.training_data import TrainingData
+from rasa.train import train_nlu
+from rasa.nlu.model import Interpreter
+from rasa.nlu.test import (
+    create_intent_report,
+    get_eval_data,
+    remove_pretrained_extractors,
+    get_intent_errors,
+)
+import rasa.utils.plotting
+
+
+def collect_intents_for_data_augmentation(
+    nlu_training_data: TrainingData,
+    num_intents_to_augment: int,
+    classification_report: Dict[Text, Dict[Text, Any]],
+) -> Set[Text]:
+    """Collects intents for which to perform data augmentation.
+
+    It analyses the training datasets and extracts:
+        * The `num_intents_to_augment` intents with the least data
+        * The `num_intents_to_augment` with lowest precision (according to `classification_report`)
+        * The `num_intents_to_augment` with lowest recall (according to `classification_report`)
+        * The `num_intents_to_augment` with lowest f1-score (according to `classification_report`)
+        * The `num_intents_to_augment` most frequently confused intents (according to `classification_report`)
+
+    For all of the intents matching the above criteria paraphrases will be used for data augmentation.
+
+    Args:
+        nlu_training_data: The existing NLU training data.
+        num_intents_to_augment: The number of intents to pick per criterion.
+        classification_report: An existing classification report (without data augmentation).
+
+    Returns:
+        The set of intent names for which data augmentation will be performed.
+    """
+    # Determine low data, low performing and frequently confused intents
+    low_data_intents = sorted(
+        nlu_training_data.number_of_examples_per_intent.items(),
+        key=operator.itemgetter(1),
+    )[:num_intents_to_augment]
+    low_precision_intents = sorted(
+        [
+            (intent, classification_report[intent]["precision"])
+            for intent in nlu_training_data.intents
+        ],
+        key=operator.itemgetter(1),
+    )[:num_intents_to_augment]
+    low_recall_intents = sorted(
+        [
+            (intent, classification_report[intent]["recall"])
+            for intent in nlu_training_data.intents
+        ],
+        key=operator.itemgetter(1),
+    )[:num_intents_to_augment]
+    low_f1_intents = sorted(
+        [
+            (intent, classification_report[intent]["f1-score"])
+            for intent in nlu_training_data.intents
+        ],
+        key=operator.itemgetter(1),
+    )[:num_intents_to_augment]
+    freq_confused_intents = sorted(
+        [
+            (intent, sum(classification_report[intent]["confused_with"].values()))
+            for intent in nlu_training_data.intents
+        ],
+        key=operator.itemgetter(1),
+        reverse=True,
+    )[:num_intents_to_augment]
```

Comment:
Do we need this if we are already accounting for low recall? Because intents with low recall will have most confusions too.

Reply:
As discussed in our call today, I've created a follow-up issue for a more thorough investigation of this in #8009. I'll also leave it in for now.